src/tests/CullingTest.cpp

```diff
@@ -0,0 +1,231 @@
+  renderCamera.node().setTransformation(cameraNode.absoluteTransformation());
+
+  // collect all the drawables and their transformations
+  std::vector<std::pair<std::reference_wrapper<Mn::SceneGraph::Drawable3D>,
+                        Mn::Matrix4>>
+      drawableTransforms = renderCamera.drawableTransformations(drawables);
+
+  // do the culling (to create the testing group)
+  std::vector<std::pair<std::reference_wrapper<Mn::SceneGraph::Drawable3D>,
+                        Mn::Matrix4>>::iterator newEndIter =
+      renderCamera.cull(drawableTransforms);
+
+  // ============== Test 1 ==================
+  // draw all the invisibles reported by cull()
+  // check passed if 0 sample is returned (that means they are indeed
+  // invisibles.)
+  {
+    // objects will contain all the invisible ones
+    std::vector<std::pair<std::reference_wrapper<Mn::SceneGraph::Drawable3D>,
+                          Mn::Matrix4>>
+        objects = renderCamera.drawableTransformations(drawables);
+
+    // CAREFUL: all the invisible ones are NOT stored at [newEndIter,
+    // drawableTransforms.end()] This is because std::remove_if will only move
+    // elements, not swap elements
+    objects.erase(
+        std::remove_if(
+            objects.begin(), objects.end(),
+            [&](const std::pair<
+                std::reference_wrapper<Mn::SceneGraph::Drawable3D>,
+                Mn::Matrix4>& a) {
+              for (std::vector<std::pair<
+                       std::reference_wrapper<Mn::SceneGraph::Drawable3D>,
+                       Mn::Matrix4>>::iterator iter =
+                       drawableTransforms.begin();
+                   iter != newEndIter; ++iter) {
+                if (std::addressof(a.first.get()) ==
+                    std::addressof(iter->first.get())) {
+                  return true;  // remove it
+                }
+              }
+              return false;
+            }),
+        objects.end());
+
+    Mn::GL::SampleQuery q{Mn::GL::SampleQuery::Target::AnySamplesPassed};
+    q.begin();
+    renderCamera.MagnumCamera::draw(objects);
+    q.end();
+
+    EXPECT_EQ(q.result<bool>(), false);
+  }
+
+  // ============== Test 2 ==================
+  // draw the visibles one by one.
+  // check if each one is a genuine visible drawable
+  unsigned int numVisiblesGroundTruth = 0;
+  auto renderOneDrawable =
+      [&](const std::pair<std::reference_wrapper<Mn::SceneGraph::Drawable3D>,
+                          Mn::Matrix4>& a) {
+        std::vector<std::pair<
+            std::reference_wrapper<Mn::SceneGraph::Drawable3D>, Mn::Matrix4>>
+            objects;
+        objects.emplace_back(a);
+
+        Mn::GL::SampleQuery q{Mn::GL::SampleQuery::Target::AnySamplesPassed};
+        q.begin();
+        renderCamera.MagnumCamera::draw(objects);
+        q.end();
+
+        // check if it a genuine visible drawable
+        EXPECT_EQ(q.result<bool>(), true);
```

Comment:
@mosra : After I fixed Test 1, now the Test 2 will fail. 

I copied the this piece of code (`renderOneDrawable` and the `for_each`) to viewer. It can render these objects on the screen, which means they are visible.

Not sure why it reports false. Did I setup anything wrong?

Reply:
When fixing test 1, I was working on a Fedora system (linux). I am pretty sure if something is rendered, the sampleQuery can capture it.

After I went home, I was using may Macbook to fix Test 2, which gave me the above error (sampleQuery told me some objects (supposed to be visible) were invisible.) I hope it would again a OS-specific issue. (Hope it is not as I can see the CI test also failed.)

I will try it tomorrow in the office with a linux machine.

Reply:
I just remotely connected to a linux box and tested it. It failed. So it should not be an OS issue.

Reply:
Random shoot -- I think you'll need a framebuffer clear before each, otherwise the depth buffer will prevent things from being rendered in subsequent runs (which iteration fails? the first one? or the subsequent ones?).

Reply:
All of them failed.
Actually, if I disable the Test 1, it is the same. (If this is the dirty framebuffer issue, at least it should succeed at the 1st one.)
I hope you can download and try this branch on your linux machine.

Also, how can I clean the framebuffer at this point? Any docs can help me? Thanks a lot!

Reply:
Depending on the OS, the depth buffer might be by default all 0s (in which case the first render would work) or all 1s (in which case it wouldn't). Ran into this exact issue on a different project last week.

How? Not sure, I mean there has to be some framebuffer the simulator had set up, no? :)

Reply:
I cannot remember where we set the framebuffer. Are we using the default framebuffer?
Anyway, I have to go. It is 2am. Hope you can try this test.

Thank you very much for all the feedbacks. I will continue tomorrow.

Reply:
I'll try to test, but not sure if today.

The default framebuffer can't really be relied on, since some platforms (macOS) don't even have it and EGL has only a 16x16 one (making all sample queries return 0 anyway).

If getting the framebuffer out of the simulator would be too painful, it's always possible to create a temporary one, with just a color buffer for example.

Reply:
@bigbike This is where we get the framebuffer: https://aihabitat.org/docs/habitat-cpp/classesp_1_1gfx_1_1RenderTarget.html

Our windowless context does not create a framebuffer, so IDK what you are even rendering too currently :)

Reply:
Thank you very much. That makes a lot of sense.
On Fedora, I can build in GUI mode. So all the tests work perfectly. Just tried.

But on CI, and devserver, they must be in headless mode, which causes failures. I need this render target.
Thank you very much, @erikwijmans !

Reply:
All right. It is indeed related to the frame buffer, and OS-specific.
After I added the RenderTarget, it works on mac OSX, Ubuntu now. (Fedora does not have this issue since the beginning.)

@mosra , @erikwijmans , your insights really help here. Thank you!